### ISSUE TITLE
lfs.fetchinclude no longer works with directories

### ISSUE THREAD

**Issue by nuxi** [git-lfs/git-lfs#4945]
**Describe the bug**

Using `lfs.fetchinclude` with a directory no longer works with the 3.x releases of git-lfs. (It worked fine with the 2.x releases)

**To Reproduce**
1. `git clone https://github.com/nuxi/git-lfs-bug.git`
2. Check the contents of `.lfsconfig` and see that `lfs.fetchinclude = foo` is set.
3. Check to see if `foo/should_be_a_binary.bin` is the pointer or a binary.
4. Check to see if `bar/should_be_a_pointer.bin` is the pointer or a binary.

**Expected behavior**

I think the filenames are obvious enough, but with git-lfs 3.x, the `foo/should_be_a_binary.bin` file is still a pointer. Under git-lfs 2.x, the file is the binary.

**System environment**

Debian Sid

**Output of `git lfs env`**

```
git-lfs/3.0.2 (GitHub; linux amd64; go 1.17.2)
git version 2.35.1

Endpoint=https://github.com/nuxi/git-lfs-bug.git/info/lfs (auth=none)
LocalWorkingDir=/home/nuxi/src/git-lfs-bug/lfs3-main
LocalGitDir=/home/nuxi/src/git-lfs-bug/lfs3-main/.git
LocalGitStorageDir=/home/nuxi/src/git-lfs-bug/lfs3-main/.git
LocalMediaDir=/home/nuxi/src/git-lfs-bug/lfs3-main/.git/lfs/objects
LocalReferenceDirs=
TempDir=/home/nuxi/src/git-lfs-bug/lfs3-main/.git/lfs/tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=/home/nuxi/src/git-lfs-bug/lfs3-main/.git/lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
FetchInclude=foo
GIT_EXEC_PATH=/usr/lib/git-core
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Additional context**

The same repo, but with the config option pointing directly at the exact binary file is fine. You can try this with `git clone -b exact https://github.com/nuxi/git-lfs-bug.git`

Support for specifying a directory is documented in the manual as the very first example - https://github.com/git-lfs/git-lfs/blob/v3.1.2/docs/man/git-lfs-fetch.1.ronn#examples

**Comment by chrisd8088** [comment]
Hey, thanks for the report and reproduction case.  It seems as though we may have converted some commands to use `gitattributes(5)`-style matching when they should be using `gitignore(5)`-style matching, back in #4556.

In addition to the documentation you mention there's some related description in the `git-lfs-pull(1)` [page](https://github.com/git-lfs/git-lfs/blob/2fee8da0d1d12d72efbe4628a5648180eed92f5a/docs/man/git-lfs-pull.1.ronn#include-and-exclude) and the `git-lfs-fetch(1)` [page](https://github.com/git-lfs/git-lfs/blob/2fee8da0d1d12d72efbe4628a5648180eed92f5a/docs/man/git-lfs-fetch.1.ronn#include-and-exclude).

Those commands [use](https://github.com/git-lfs/git-lfs/blob/2fee8da0d1d12d72efbe4628a5648180eed92f5a/commands/command_fetch.go#L92) the `buildFilepathFilter()` function which [passes](https://github.com/git-lfs/git-lfs/blob/2fee8da0d1d12d72efbe4628a5648180eed92f5a/commands/commands.go#L133) its arguments, including any `-I`/`--include` and `-X`/`--exclude` command-line options, [down](https://github.com/git-lfs/git-lfs/blob/2fee8da0d1d12d72efbe4628a5648180eed92f5a/commands/commands.go#L137) to `determineIncludeExcludePaths()` which returns those if they exist, and otherwise any `lfs.fetchinclude` [setting](https://github.com/git-lfs/git-lfs/blob/2fee8da0d1d12d72efbe4628a5648180eed92f5a/config/config.go#L195) or `lfs.fetchexclude` [setting](https://github.com/git-lfs/git-lfs/blob/2fee8da0d1d12d72efbe4628a5648180eed92f5a/config/config.go#L200).  And `buildFilepathFilter()` [specifies](https://github.com/git-lfs/git-lfs/blob/2fee8da0d1d12d72efbe4628a5648180eed92f5a/commands/commands.go#L133) `gitignore(5)`-style matching, which allows for a directory match to recursively match everything within it.  So far, so good.

When you perform a `git clone`, however, those Git LFS commands aren't used; instead, `git-lfs-smudge(1)` or, in the efficient case, `git-lfs-filter-process(1)` will be run by Git.  And for those, since #4556, we [interpret](https://github.com/git-lfs/git-lfs/blob/2fee8da0d1d12d72efbe4628a5648180eed92f5a/commands/command_smudge.go#L157) the same `lfs.fetchinclude` and `lfs.fetchexclude` configuration values `gitattributes(5)`-style matching.

I should be able to have a PR to revise this ready fairly soon, but I think it would be good to add some tests (because right now it looks like the test suite only covers the use of those configuration settings with `git-lfs-clone`, `git-lfs-fetch`, and `git-lfs-pull`, which is probably why we missed the others), and tune up the documentation a bit as well.  So it might take a short while for me to get those all together.

Thanks again for the report!

**Comment by chrisd8088** [comment]
I notice we also have another thing to consider, which is that `git lfs migrate` wants to treat its `-I`/`-X` command line arguments using `gitattributes(5)`-style matching; that was the point of #4758 which restored some pre-3.0 behaviour for that as well, again stemming from the wildmatch fixes in #4556.

However, that means we're treating any `lfs.fetchinclude` or `lfs.fetchexclude` configurations using that style of matching as well, if they are specified and `git lfs migrate` is run.  I think we'll probably have to decide to treat that as an unsupported use case, and stop using them as input for how migration operates at all.  That may catch a few people off guard; I'm not sure if that's a breaking change or just a bug fix.

But this may take some thinking, because clearly we need to treat the `lfs.fetchinclude` and `lfs.fetchexclude` settings the same way in all cases.
